// @terra-money/use-wallet/useConnectedWallet.js
import { AccAddress } from '@terra-money/terra.js';
import { useMemo } from 'react';
import { WalletStatus, } from './types';
import { useWallet } from './useWallet';
export function useConnectedWallet() {
    const { status, network, wallets, post, sign, signBytes, supportFeatures } = useWallet();
    const value = useMemo(() => {
        try {
            if (status === WalletStatus.WALLET_CONNECTED &&
                wallets.length > 0 &&
                AccAddress.validate(wallets[0].terraAddress)) {
                const { terraAddress, connectType, design } = wallets[0];
                return {
                    network,
                    terraAddress: terraAddress,
                    walletAddress: terraAddress,
                    design,
                    post: (tx) => {
                        return post(tx, terraAddress);
                    },
                    sign: (tx) => {
                        return sign(tx, terraAddress);
                    },
                    signBytes: (bytes) => {
                        return signBytes(bytes, terraAddress);
                    },
                    availablePost: supportFeatures.has('post'),
                    availableSign: supportFeatures.has('sign'),
                    availableSignBytes: supportFeatures.has('sign-bytes'),
                    connectType,
                };
            }
            else {
                return undefined;
            }
        }
        catch (_a) {
            return undefined;
        }
    }, [network, post, sign, signBytes, status, supportFeatures, wallets]);
    return value;
}
