// @terra-money/wallet-provider/modules/extension-router/defaultSelectModal.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { fixHMR } from 'fix-hmr';
import { createElement } from 'react';
import { render } from 'react-dom';
import styled, { keyframes } from 'styled-components';
export function defaultSelectModal(extensionInfos) {
    return new Promise((resolve) => {
        var _a;
        const modalContainer = window.document.createElement('div');
        function onComplete(extensionInfo) {
            var _a;
            resolve(extensionInfo);
            (_a = modalContainer.parentElement) === null || _a === void 0 ? void 0 : _a.removeChild(modalContainer);
        }
        const modal = createElement(Modal, {
            extensionInfos,
            onComplete,
        });
        render(modal, modalContainer);
        (_a = window.document.querySelector('body')) === null || _a === void 0 ? void 0 : _a.appendChild(modalContainer);
    });
}
function Component({ className, extensionInfos, onComplete, }) {
    return (_jsxs("div", Object.assign({ className: className }, { children: [_jsx("div", { onClick: () => onComplete(null) }, void 0), _jsxs("section", { children: [_jsx("h1", { children: "Select a wallet" }, void 0), _jsx("ul", { children: extensionInfos.map((extensionInfo) => (_jsx("li", { children: _jsxs("button", Object.assign({ onClick: () => onComplete(extensionInfo) }, { children: [_jsx("img", { src: extensionInfo.icon, alt: extensionInfo.name }, void 0), _jsxs("span", { children: [extensionInfo.name, " [", extensionInfo.identifier, "]"] }, void 0)] }), void 0) }, extensionInfo.identifier))) }, void 0)] }, void 0)] }), void 0));
}
const modalEnter = keyframes `
  0% {
    opacity: 0;
  }
  
  100% {
    opacity: 1;
  }
`;
const sectionEnter = keyframes `
  0% {
    opacity: 0;
    transform: scale(0.4);
  }
  
  100% {
    opacity: 1;
    transform: scale(1);
  }
`;
const StyledComponent = styled(Component) `
  position: fixed;
  z-index: 100000;

  color: #000000;

  left: 0;
  top: 0;
  width: 100vw;
  height: 100vh;

  display: grid;
  place-content: center;

  > div {
    position: fixed;
    z-index: -1;

    left: 0;
    top: 0;
    width: 100vw;
    height: 100vh;
    background-color: rgba(0, 0, 0, 0.3);

    animation: ${modalEnter} 0.2s ease-in-out;
  }

  > section {
    border-radius: 25px;

    background-color: #ffffff;
    box-shadow: 0 4px 18px 3px rgba(0, 0, 0, 0.43);

    animation: ${sectionEnter} 0.2s ease-in-out;

    padding: 20px 30px;

    h1 {
      font-size: 27px;
      font-weight: 500;

      text-align: center;

      margin-bottom: 24px;
    }

    ul {
      padding: 0;
      list-style: none;

      display: flex;
      flex-direction: column;
      gap: 10px;

      button {
        border: none;
        background-color: transparent;
        outline: none;
        cursor: pointer;

        display: flex;
        gap: 10px;
        align-items: center;

        img {
          width: 30px;
          height: 30px;
        }

        span {
          font-size: 12px;
        }
      }
    }
  }
`;
const Modal = fixHMR(StyledComponent);
