// @terra-money/wallet-provider/impl-react/StaticWalletProvider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { WalletContext, WalletStatus, } from '@terra-money/use-wallet';
import { useMemo } from 'react';
export function StaticWalletProvider({ children, defaultNetwork, status = WalletStatus.INITIALIZING, availableConnectTypes = [], availableInstallTypes = [], availableConnections = [], wallets = [], supportFeatures = new Set(), }) {
    const state = useMemo(() => {
        return {
            availableConnectTypes,
            availableInstallTypes,
            availableConnections,
            status,
            network: defaultNetwork,
            wallets,
            supportFeatures,
            install: () => {
                throw new Error('not implemented!');
            },
            connect: () => {
                throw new Error('not implemented!');
            },
            connectReadonly: () => {
                throw new Error('not implemented!');
            },
            disconnect: () => {
                throw new Error('not implemented!');
            },
            post: () => {
                throw new Error('not implemented!');
            },
            sign: () => {
                throw new Error('not implemented!');
            },
            signBytes: () => {
                throw new Error('not implemented!');
            },
            refetchStates: () => {
                throw new Error('not implemented!');
            },
            recheckStatus: () => {
                throw new Error('not implemented!');
            },
            isChromeExtensionCompatibleBrowser: () => {
                throw new Error('not implemented!');
            },
            hasCW20Tokens: () => {
                throw new Error('not implemented!');
            },
            addCW20Tokens: () => {
                throw new Error('not implemented!');
            },
            hasNetwork: () => {
                throw new Error('not implemented!');
            },
            addNetwork: () => {
                throw new Error('not implemented!');
            },
        };
    }, [
        availableConnectTypes,
        availableInstallTypes,
        availableConnections,
        status,
        defaultNetwork,
        wallets,
        supportFeatures,
    ]);
    return (_jsx(WalletContext.Provider, Object.assign({ value: state }, { children: children }), void 0));
}
