// @terra-money/wallet-provider/_commonjs/modules/extension-router/defaultSelectModal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultSelectModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const fix_hmr_1 = require("fix-hmr");
const react_1 = require("react");
const react_dom_1 = require("react-dom");
const styled_components_1 = __importStar(require("styled-components"));
function defaultSelectModal(extensionInfos) {
    return new Promise((resolve) => {
        var _a;
        const modalContainer = window.document.createElement('div');
        function onComplete(extensionInfo) {
            var _a;
            resolve(extensionInfo);
            (_a = modalContainer.parentElement) === null || _a === void 0 ? void 0 : _a.removeChild(modalContainer);
        }
        const modal = (0, react_1.createElement)(Modal, {
            extensionInfos,
            onComplete,
        });
        (0, react_dom_1.render)(modal, modalContainer);
        (_a = window.document.querySelector('body')) === null || _a === void 0 ? void 0 : _a.appendChild(modalContainer);
    });
}
exports.defaultSelectModal = defaultSelectModal;
function Component({ className, extensionInfos, onComplete, }) {
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ className: className }, { children: [(0, jsx_runtime_1.jsx)("div", { onClick: () => onComplete(null) }, void 0), (0, jsx_runtime_1.jsxs)("section", { children: [(0, jsx_runtime_1.jsx)("h1", { children: "Select a wallet" }, void 0), (0, jsx_runtime_1.jsx)("ul", { children: extensionInfos.map((extensionInfo) => ((0, jsx_runtime_1.jsx)("li", { children: (0, jsx_runtime_1.jsxs)("button", Object.assign({ onClick: () => onComplete(extensionInfo) }, { children: [(0, jsx_runtime_1.jsx)("img", { src: extensionInfo.icon, alt: extensionInfo.name }, void 0), (0, jsx_runtime_1.jsxs)("span", { children: [extensionInfo.name, " [", extensionInfo.identifier, "]"] }, void 0)] }), void 0) }, extensionInfo.identifier))) }, void 0)] }, void 0)] }), void 0));
}
const modalEnter = (0, styled_components_1.keyframes) `
  0% {
    opacity: 0;
  }
  
  100% {
    opacity: 1;
  }
`;
const sectionEnter = (0, styled_components_1.keyframes) `
  0% {
    opacity: 0;
    transform: scale(0.4);
  }
  
  100% {
    opacity: 1;
    transform: scale(1);
  }
`;
const StyledComponent = (0, styled_components_1.default)(Component) `
  position: fixed;
  z-index: 100000;

  color: #000000;

  left: 0;
  top: 0;
  width: 100vw;
  height: 100vh;

  display: grid;
  place-content: center;

  > div {
    position: fixed;
    z-index: -1;

    left: 0;
    top: 0;
    width: 100vw;
    height: 100vh;
    background-color: rgba(0, 0, 0, 0.3);

    animation: ${modalEnter} 0.2s ease-in-out;
  }

  > section {
    border-radius: 25px;

    background-color: #ffffff;
    box-shadow: 0 4px 18px 3px rgba(0, 0, 0, 0.43);

    animation: ${sectionEnter} 0.2s ease-in-out;

    padding: 20px 30px;

    h1 {
      font-size: 27px;
      font-weight: 500;

      text-align: center;

      margin-bottom: 24px;
    }

    ul {
      padding: 0;
      list-style: none;

      display: flex;
      flex-direction: column;
      gap: 10px;

      button {
        border: none;
        background-color: transparent;
        outline: none;
        cursor: pointer;

        display: flex;
        gap: 10px;
        align-items: center;

        img {
          width: 30px;
          height: 30px;
        }

        span {
          font-size: 12px;
        }
      }
    }
  }
`;
const Modal = (0, fix_hmr_1.fixHMR)(StyledComponent);
