// @terra-money/wallet-provider/_commonjs/modules/walletconnect/modal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraWalletconnectQrcodeModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const qrcode_react_1 = __importDefault(require("qrcode.react"));
const react_1 = require("react");
const react_dom_1 = require("react-dom");
const styled_components_1 = __importStar(require("styled-components"));
const browser_check_1 = require("../../utils/browser-check");
class TerraWalletconnectQrcodeModal {
    constructor() {
        this.modalContainer = null;
        this.callback = null;
        this.setCloseCallback = (callback) => {
            this.callback = callback;
        };
        this.open = (uri, cb, _qrcodeModalOptions) => {
            var _a;
            const modalContainer = window.document.createElement('div');
            const query = encodeURIComponent(`action=wallet_connect&payload=${encodeURIComponent(uri)}`);
            const schemeUri = `https://terrastation.page.link/?link=https://terra.money?${query}&apn=money.terra.station&ibi=money.terra.station&isi=1548434735`;
            const modal = (0, react_1.createElement)(TerraQRCodeModal, {
                schemeUri,
                onClose: () => {
                    if (this.callback) {
                        this.callback();
                        this.callback = null;
                    }
                    this.close();
                },
            });
            if ((0, browser_check_1.isMobile)()) {
                window.location.href = schemeUri;
            }
            (0, react_dom_1.render)(modal, modalContainer);
            (_a = window.document.querySelector('body')) === null || _a === void 0 ? void 0 : _a.appendChild(modalContainer);
            this.modalContainer = modalContainer;
        };
        this.close = () => {
            var _a;
            if (this.modalContainer) {
                (_a = this.modalContainer.parentElement) === null || _a === void 0 ? void 0 : _a.removeChild(this.modalContainer);
            }
            this.callback = null;
        };
    }
}
exports.TerraWalletconnectQrcodeModal = TerraWalletconnectQrcodeModal;
function TerraQRCodeModalBase({ schemeUri, className, onClose, }) {
    const isMobile = (0, react_1.useMemo)(() => (0, browser_check_1.isMobile)(), []);
    const openTerraStationMobile = (0, react_1.useCallback)(() => {
        window.location.href = schemeUri;
    }, [schemeUri]);
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ className: className }, { children: [(0, jsx_runtime_1.jsx)("div", { onClick: onClose }, void 0), isMobile ? ((0, jsx_runtime_1.jsxs)("section", Object.assign({ className: "mobile" }, { children: [(0, jsx_runtime_1.jsx)("h1", { children: "Wallet Connect" }, void 0), (0, jsx_runtime_1.jsx)("button", Object.assign({ onClick: openTerraStationMobile, className: "flat-button" }, { children: "Open Terra Station Mobile" }), void 0)] }), void 0)) : ((0, jsx_runtime_1.jsxs)("section", Object.assign({ className: "desktop" }, { children: [(0, jsx_runtime_1.jsx)("h1", { children: "Wallet Connect" }, void 0), (0, jsx_runtime_1.jsx)(qrcode_react_1.default, { value: schemeUri, size: 240 }, void 0)] }), void 0))] }), void 0));
}
const modalEnter = (0, styled_components_1.keyframes) `
  0% {
    opacity: 0;
  }
  
  100% {
    opacity: 1;
  }
`;
const sectionEnter = (0, styled_components_1.keyframes) `
  0% {
    opacity: 0;
    transform: scale(0.4);
  }
  
  100% {
    opacity: 1;
    transform: scale(1);
  }
`;
const TerraQRCodeModal = (0, styled_components_1.default)(TerraQRCodeModalBase) `
  position: fixed;
  z-index: 100000;

  color: #000000;

  left: 0;
  top: 0;
  width: 100vw;
  height: 100vh;

  display: grid;
  place-content: center;

  > div {
    position: fixed;
    z-index: -1;

    left: 0;
    top: 0;
    width: 100vw;
    height: 100vh;
    background-color: rgba(0, 0, 0, 0.3);

    animation: ${modalEnter} 0.2s ease-in-out;
  }

  > section {
    border-radius: 25px;

    background-color: #ffffff;
    box-shadow: 0 4px 18px 3px rgba(0, 0, 0, 0.43);

    animation: ${sectionEnter} 0.2s ease-in-out;

    button {
      cursor: pointer;

      display: block;
      outline: none;
      background-color: transparent;
      width: 100%;
      height: 32px;
      font-size: 13px;
      letter-spacing: -0.2px;
      border-radius: 18px;
      border: solid 1px #2c2c2c;

      &.flat-button {
        border: 0;
        color: #ffffff;
        background-color: #2c2c2c;
      }
    }

    .separator {
      height: 12px;

      position: relative;

      hr {
        position: absolute;
        top: 5px;
        left: 0;
        right: 0;

        border: 0;
        border-bottom: 1px dashed #cccccc;
      }

      span {
        display: block;

        position: absolute;
        top: -4px;
        left: 50%;
        transform: translateX(-50%);

        font-size: 12px;
        color: #c2c2c2;
        background-color: #ffffff;
        padding: 0 2px;
      }
    }

    &.desktop {
      padding: 50px 60px;

      h1 {
        font-size: 27px;
        font-weight: 500;

        text-align: center;

        margin-bottom: 24px;
      }

      button {
        margin-top: 20px;
      }
    }

    &.mobile {
      padding: 40px 30px;
      min-width: 320px;

      h1 {
        font-size: 22px;
        font-weight: 500;

        text-align: center;

        margin-bottom: 30px;
      }

      .separator {
        margin: 10px 0;
      }

      canvas {
        display: block;
        margin: 16px auto 0 auto;
      }
    }
  }
`;
